### PR TITLE
mem-cache: Unit test FIFO RP

### PR DESCRIPTION
**Pull request by odanrc**
Add unit tests for the FIFO replacement policy.

As side effects:
    - We now verify that the repl data is not null
    - Fixed a bug where invalidated entries would not be prioritized during the victimization process

### CODE REVIEW

**Comment by odanrc** [comment]
@BobbyRBruce locally I have added tests for MRU, FIFO, LFU, LRU and 2nd chance. So far only FIFO had a bug.

**Comment by powerjg** [comment]
So cool that you found a bug! The failing test is just a runner being flakey. We can re-run once the other tests pass.

**Comment by odanrc** [comment]
> So cool that you found a bug!

Yeah! Now that we can easily add tests to sim objects we should be able to spot more (or hopefully none :)).

> The failing test is just a runner being flakey. We can re-run once the other tests pass.

In this case it was because I reordered my commits and forgot to pick one of them. It should pass now.

File: src/mem/cache/replacement_policies/fifo_rp.test.cc [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    rp->touch(entry.replacementData);
+    ASSERT_EQ(rp->getVictim(candidates), &entry);
+}
+
+/// Fixture that tests victimization
+class FIFORPVictimizationTestF : public FIFORPTestF
+{
+  protected:
+    // The entries being victimized
+    std::vector<gem5::ReplaceableEntry> entries;
```
Comment (giactra) [inline src/mem/cache/replacement_policies/fifo_rp.test.cc:85]: Do we really need the entries vector? Can't we just use the candidates one?

File: src/mem/cache/replacement_policies/fifo_rp.test.cc [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            candidates.push_back(&entry);
+        }
+    }
+};
+
+/// Test that when all entries are invalid a single entry will always be
+/// selected, regardless of the order of the invalidations
+TEST_F(FIFORPVictimizationTestF, GetVictimAllInvalid)
+{
+    auto expected_victim = &entries[0];
```
Comment (giactra) [inline src/mem/cache/replacement_policies/fifo_rp.test.cc]: Nit: "front()" seems like a more descriptive way of referencing the first entry

File: src/mem/cache/replacement_policies/fifo_rp.test.cc [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    // having more entries is not expected to increase coverage
+    FIFORPVictimizationTestF() : FIFORPTestF(), entries(4) {
+        for (auto &entry : entries) {
+            entry.replacementData = rp->instantiateEntry();
+            candidates.push_back(&entry);
+        }
+    }
+};
+
+/// Test that when all entries are invalid a single entry will always be
```
Comment (giactra) [inline src/mem/cache/replacement_policies/fifo_rp.test.cc]: What do you mean by a single entry?
Do you mean the first one to be instantiated? It's not clear from the test description which candidate will be selected in that scenario

File: src/mem/cache/replacement_policies/fifo_rp.test.cc [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    // Since all candidates are already invalid, nothing changes if we
+    // invalidate all of them again
+    for (auto &entry : entries) {
+        rp->invalidate(entry.replacementData);
+    }
+    ASSERT_EQ(rp->getVictim(candidates), expected_victim);
+
+    // Even if we invalidate the entry being selected for victimization last
+    rp->invalidate(expected_victim->replacementData);
+    ASSERT_EQ(rp->getVictim(candidates), expected_victim);
```
Comment (giactra) [inline src/mem/cache/replacement_policies/fifo_rp.test.cc]: Considering the scope of the test is to check:

```
/// Test that when all entries are invalid a single entry will always be
/// selected, regardless of the order of the invalidations
```

I believe a simpler fixture generation would construct the test with all entries in a VALID state. So you can have a simpler test case:

    // Reverse invalidation order
    for (auto it = candidates.rbegin(); it != candidates.rend(); ++it) {
        rp->invalidate(it->replacementData);
    }
    ASSERT_EQ(rp->getVictim(candidates), expected_victim);

**Review by giactra — COMMENTED** [review]
That is amazing @odanrc !
Thanks for this valuable contribution! Looking forward to see some tests for LRU and RRIP! I have some comments wrt the unit tests

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 1 above)
Comment (odanrc) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc:85]: Candidates is a vector of pointers, so we need to instantiate them somewhere, which is what entries is being used for.

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 3 above)
Comment (odanrc) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc]: I will try to clarify the comment, but I mean that it always picks the first entry in the candidate list.

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 4 above)
Comment (odanrc) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc]: For context, although I think you understood that, the name of this test is GetVictimAllInvalid, so its goal is to determine which entry is picked when all entries are invalid. The following test, GetVictimOneInvalid, determines the behavior when all entries but one are valid. Finally, GetVictim determines the behavior when all entries are valid.

Regarding your suggestion, I actually added lines 143 and 142 on purpose to reduce the likelihood of some kind of bias happening inside the invalidation function. For example, under your proposal, what guarantees that the victim is front() because it was the last to be invalidated, and not because it is the first candidate? I wanted to 'make sure' that no matter the order at which an invalidation happens, the victim is deterministic (front()), if all candidates are invalid.

**Comment by odanrc** [comment]
I only changed 2 lines in the last upload, so I squashed into the first commit. Here are the changes:

```
-/// Test that when all entries are invalid a single entry will always be
+/// Test that when all entries are invalid the first candidate will always be
 /// selected, regardless of the order of the invalidations
 TEST_F(FIFORPVictimizationTestF, GetVictimAllInvalid)
 {
-    auto expected_victim = &entries[0];
+    auto expected_victim = &entries.front();
```

**Comment by odanrc** [comment]
@giactra , have you got some time to review this again?

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 4 above)
Comment (giactra) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc]: > For example, under your proposal, what guarantees that the victim is front() because it was the last to be invalidated, and not because it is the first candidate?

I am not sure I understand. I could ask the same about the current version?

There's really no reason why invalidating outside of a loop is better than within a loop. Just doing

```
// Reverse invalidation order
for (auto it = candidates.rbegin(); it != candidates.rend(); ++it) {
    rp->invalidate(it->replacementData);
}
```

Will guarantee the expected victim will be invalidated last, Similarly to what you do by adding an out of the loop invalidation call.
I believe we should try to make unit-tests as simple as possible and avoid calling too many methods. If you fear there is a risk of an entry being selected because it is the first of the candidates, we should add a separate test that solely checks for that

File: src/mem/cache/replacement_policies/fifo_rp.cc [hunk 5]
```diff
@@ -48,20 +48,27 @@ FIFO::FIFO(const Params &p)
...
 FIFO::invalidate(const std::shared_ptr<ReplacementData>& replacement_data)
 {
-    // Reset insertion tick
+    assert(replacement_data);
+
+    // To prioritize these entries to be selected during victimization
+    // we set their insertion tick as 0
     std::static_pointer_cast<FIFOReplData>(
-        replacement_data)->tickInserted = ++timeTicks;
+        replacement_data)->tickInserted = 0;
```
Comment (giactra) [inline src/mem/cache/replacement_policies/fifo_rp.cc]: This is the bug fix. Shall we push this as a separate commit?

**Review by giactra — CHANGES_REQUESTED** [review]
Hi @odanrc apologies for the late review. 

Wrt to the first commit which is making the ReplacementData polymorphic: 

> This struct is supposed to be used polymorphically, so make
> it possible to do so.

IMHO this is not true. ReplacementData is simply an empty struct whose sole purpose is to serve as an interface for polymophic ReplacementPolicy. There is really no virtual method defined in ReplacementData so I fail to understand how it can be considered polymorphic.

I understand adding a vtable is useful for dynamic type checking in your unit-tests, but I don't consider this to be a sufficient condition for adding a virtual method to the class. Let me know if you disagree. If there is consensus on paying this penalty, I believe it should be at least explicit.
I fear though that transforming ReplacementData into a poly type might make the design of replacement framework much more complicated (but you are the expert here so please correct me if I am wrong)

File: src/mem/cache/replacement_policies/fifo_rp.cc (hunk 5 above)
Comment (odanrc) [reply src/mem/cache/replacement_policies/fifo_rp.cc]: Done: https://github.com/gem5/gem5/pull/2412

**Comment by odanrc** [comment]
> Hi @odanrc apologies for the late review.
> 
> Wrt to the first commit which is making the ReplacementData polymorphic:
> 
> > This struct is supposed to be used polymorphically, so make
> > it possible to do so.
> 
> IMHO this is not true. ReplacementData is simply an empty struct whose sole purpose is to serve as an interface for polymophic ReplacementPolicy. There is really no virtual method defined in ReplacementData so I fail to understand how it can be considered polymorphic.
> 
> I understand adding a vtable is useful for dynamic type checking in your unit-tests, but I don't consider this to be a sufficient condition for adding a virtual method to the class. Let me know if you disagree. If there is consensus on paying this penalty, I believe it should be at least explicit. I fear though that transforming ReplacementData into a poly type might make the design of replacement framework much more complicated (but you are the expert here so please correct me if I am wrong)

Turns out I made the destructor virtual because of a compilation error (using dynamic cast) from the test. I fixed that by reducing the test coverage, as we shouldn´t be modifying the source to accommodate for a test.

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 4 above)
Comment (powerjg) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc]: I don't think it hurts to "double check" this. At this point, I don't think we want to stall this PR if we don't have to.

**Review by powerjg — COMMENTED** [review]
@odanrc is this ready for review?

File: src/mem/cache/replacement_policies/fifo_rp.test.cc (hunk 4 above)
Comment (odanrc) [reply src/mem/cache/replacement_policies/fifo_rp.test.cc]: I seem to have not seen Giacomo's latest comment. I have just updated the code as he suggested (last commit).

**Review by powerjg — DISMISSED** [review]

**Comment by powerjg** [comment]
@odanrc It looks like the test is failing. Other than that, the code looks good to me.

**Comment by odanrc** [comment]
> @odanrc It looks like the test is failing. Other than that, the code looks good to me.

Solved it. It was a case of forgetting to dereference an iterator.

**Comment by odanrc** [comment]
@powerjg @giactra @erin-le @BobbyRBruce  , could someone merge this PR? I have been getting 1-2 automated e-mails about it *every single day* since its inception. I really don't know why - they claim that this PR fails the CI, but... it doesn't?

**Review by BobbyRBruce — APPROVED** [review]
